pkg/events/default.go

```diff
@@ -14,24 +14,60 @@
 
 package events
 
-import "context"
+import (
+	"context"
+	"sync"
+)
 
-// DefaultPubSub is the default event pubsub.
-var DefaultPubSub = NewPubSub(DefaultBufferSize)
+var (
+	defaultPubSubMu = &sync.Mutex{}
+	// defaultPubSub is the default event pubsub.
+	defaultPubSub = NewPubSub(DefaultBufferSize)
+	// defaultPubSubSet records whether SetDefaultPubSub has been called or not.
+	defaultPubSubSet bool
+)
+
+// SetDefaultPubSub sets pubsub used by the package to ps and returns a function,
+// which can be used to undo the action.
+// SetDefaultPubSub panics if called repeatedly before undo function is called.
+func SetDefaultPubSub(ps PubSub) func() {
+	defaultPubSubMu.Lock()
+	if defaultPubSubSet {
+		panic("SetDefaultPubSub called before previous usage finished")
```

Comment:
I still think this is weird. The intention of `SetDefaultPubSub` is to replace the pubsub implementation. In production code there is absolutely no requirement of temporarily replacing it and blocking any other replacements until "finished". I understand that this can be useful for testing, but then this belongs in test code (or one of our test packages), and not in production code.

Why don't we move this `defaultPubSubSet` checking to some kind of `TemporarilyReplaceDefaultPubSub` in `pkg/util/test/events.go`?

Reply:
Another (imo cleaner) option is having the `events.{Publish,Subscribe,Unsubscribe}` funcs take a `context.Context`. Then we can implement a `events.NewContextWithPubSub` and `events.PubSubFromContext` (with fallback to `DefaultPubSub`)